Take DST into account when locating TimeZone from Numeric

yasyf opened the pull request:
When given a specific offset, use the first result found where the
total current offset (including any periodic deviations such as DST)
from UTC is equal.

Fixes #15209.

eileencodes commented:
@yasyf can you please add a test for this change? Thanks! 

When you change the test instead of adding a new commit you can do:
`git commit --amend`
`git push -f origin your-branch`

You'll also need to rebase this change because there is a conflict. Most likely the conflict is in the CHANGELOG. Let me know if you don't know how to rebase.

yasyf commented:
@eileencodes added a test, and fixed the conflict!

yasyf committed:
Take DST into account when locating TimeZone from Numeric.

When given a specific offset, use the first result found where the
total current offset (including any periodic deviations such as DST)
from UTC is equal.

rafaelfranca commented:
Backported at e4354579496339691ff15f61d228e454c529c39f

pixeltrix commented:
So I think we need to revert this as it cause a problem because it's not consistent throughout the year. For example during the winter you will get the Alaska timezone if use -9 but an error during the summer as there is no other timezone with -9. I think this discrepancy will likely confuse developers (tbh I think searching on numeric offsets is weird anyway) so unless you have any other comments @yasyf we'll revert it.

@rafaelfranca even if we don't revert it on master it should be reverted on 4-2-stable as it's a breaking change - agreed?

eileencodes commented:
For reference the errors can be seen in the most recent build. There are a few errors/failures in ActiveRecord and some in ActiveModel. This is a perfect example of why programmers hate timezones/DST :grin:, no idea there's a problem until the time changes :wink:.

I pasted one of the errors below:

```
  2) Error:
AttributeMethodsTest#test_setting_time_zone_aware_attribute_with_string:
ArgumentError: Invalid Timezone: -9
    /home/travis/build/rails/rails/activesupport/lib/active_support/core_ext/time/zones.rb:81:in `rescue in find_zone!'
    /home/travis/build/rails/rails/activesupport/lib/active_support/core_ext/time/zones.rb:65:in `find_zone!'
    /home/travis/build/rails/rails/activesupport/lib/active_support/core_ext/date_and_time/zones.rb:20:in `in_time_zone'
    /home/travis/build/rails/rails/activerecord/test/cases/attribute_methods_test.rb:617:in `block in test_setting_time_zone_aware_attribute_with_string'
    /home/travis/build/rails/rails/activesupport/lib/active_support/core_ext/range/each.rb:7:in `each'
    /home/travis/build/rails/rails/activesupport/lib/active_support/core_ext/range/each.rb:7:in `each_with_time_with_zone'
    /home/travis/build/rails/rails/activerecord/test/cases/attribute_methods_test.rb:616:in `test_setting_time_zone_aware_attribute_with_string'
```

matthewd commented:
I agree we should revert on 4-2-stable.

On master, I think we should retain the new behaviour... but instead of referring to `current_period`, we need to be working with the period that was in effect on the date in question.

pixeltrix commented:
I've reverted both master and 4-2-stable since even on master we want it to be a new API method and not via `[]`. @yasyf I've outlined what we want to do in the revert message - do you still want to work on this one?

yasyf commented:
@pixeltrix I'll take a look at the revert message and give this another try!

yasyf commented:
I have an initial draft of what you described at #19279.
